linera-core/src/client/validator_trackers.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::{BTreeMap, BTreeSet, HashMap, VecDeque};
+
+use linera_base::crypto::ValidatorPublicKey;
+use linera_chain::data_types::ChainAndHeight;
+
+/// Manages a "tracker"  of a single validator.
+/// The received log is the list of chains and heights of blocks sending messages to a
+/// particular receiver chain. The tracker is the index of the first entry in that log
+/// that corresponds to a block that hasn't been processed yet.
+/// In order to keep the tracker value up to date, we keep the part of the log starting
+/// with the first entry corresponding to a not-yet-processed block, and a record of which
+/// certificates we have already downloaded. Whenever we download a certificate corresponding to
+/// the first block in the log, we increase the tracker and pop the blocks off the log,
+/// until we hit one we haven't downloaded yet.
+pub(super) struct ValidatorTracker {
```

Comment:
Can you add the receiver's chain id into the struct? You may not need it but I think it's useful to _complete_ the struct. Otherwise, on its own, this struct is indistinguishable for another chain.